massgov/mds-licensing
# Add ingest pipeline to include OPSI data

This PR is the piece that's necessary to add OPSI data to the licensing API. It:

* Adds an additional [Python Lambda function](https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/licensing-ingest-prod-ingest/versions/$LATEST?tab=graph) to handle reaching out to the Lightsail server to pick up new data files for each agency.  This will replace the existing cron scripts that are in place on the Lightsail server. It should be fine to run both the cron scripts and the lambda at the same time - they shouldn't collide in any way. We can tear down the cron scripts once we're up and running.
* Adds ETL processing of OPSI data so that it is included in the database.
* Adds a list of agencies (`agencies.json`) that defines where we keep data for each agency, and how that agency's data should be processed.
* Changes the license details API to allow querying for OPSI licenses (`dept_name = 'OPSI'`).
* Changes the license details API to return OPSI licenses when you query for EOPSS licenses.  This is a temporary BC layer so that existing API consumers do not need to change their implementation overnight.

*Steps to Test*:
For testing, this work has been deployed to our dev environment.

* Using the attached Postman profile, create a new "Environment" for our dev environment using the following values:
    * `baseUrl`: https://dev.licensing.api.digital.mass.gov/v1
    * `apiKey`: Our standard testing API key, which is `Q8jA35Ekw785DW7jPreqedvOb1wgadx2k1xbP9De`
    * `username`: Your username (must be created at https://licensing.api.digital.mass.gov/register.html first)
    * `password`: Your password
* Use the "GetLicense" request to fetch license data on license `AI-119025` (`{{baseUrl}}/licenses/AI-119025`).  Verify that two records appear (one will be from from OPSI, one from EOPSS).
* Use the "GetLicenseByAgency" request to fetch license data for `OPSI` license `AI-119025`.  There should be one license returned, because you are querying only the OPSI dataset.
* Use the "GetLicenseByAgency" request to fetch license data for `EOPSS` license `AI-119025`.  There should be two license returned, because you are querying both the OPSI and EOPSS datasets (before we launch this change, EOPSS will drop OPSI licenses from their dataset).

To test the pipeline process:
* Invoke the [ingest lambda function](https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/licensing-ingest-prod-ingest?tab=graph) and verify that it runs without errors.
* Look into the [S3 bucket's `output` folder for OPSI](https://s3.console.aws.amazon.com/s3/buckets/api-upload.digital.mass.gov/data/output/OPSI/?region=us-east-1&tab=overview) and verify that you see a file from at least yesterday.
